### ISSUE TITLE
salt provisioner: The following settings shouldn't exist: masterless?

### ISSUE THREAD

**Issue by amaclean199** [hashicorp/vagrant#9174]
### Vagrant version
Vagrant 2.0.1

### Host operating system
macOS Sierra 10.12.6

### Guest operating system
Ubuntu 14.04

### Vagrantfile
```ruby
# -*- mode: ruby -*-
# vi: set ft=ruby ts=2 sw=2 sts=2 et :

require 'yaml'

current_dir    = File.dirname(File.expand_path(__FILE__))
configs        = YAML.load_file("#{current_dir}/config.yaml")
vsphere_config = configs['vsphere']

required_plugins = %w( vagrant-vsphere )
missing_plugins = required_plugins.reject {|p| Vagrant.has_plugin? p}
raise "vagrant plugin install #{missing_plugins.join(' ')}" unless missing_plugins.empty?

Vagrant.configure('2') do |config|

  # REDACTED site information for downloading base box

  # Pick up the site from the company-site environment variable.
  site = (ENV['company_site'] || 'place').intern
  site_settings = sites[site]
  raise "unknown site #{site}" if site_settings.nil?

  # This option disables the replacement of the insecure key pairs vagrant uses for initialisation.
  # this should only be disabled when a new box image is being created.
  # config.ssh.insert_key = false

  if site_settings.key? :proxy
    if !Vagrant.has_plugin?('vagrant-proxyconf')
        raise 'The vagrant proxy plugin is not installed.  Please install it as per the instructions on the wiki.'
    end

    # Copy all of the properties from the hash onto the config.proxy object
    site_settings[:proxy].each {|k, v| config.proxy.send("#{k}=", v)}
    puts "Using proxy url: '#{config.proxy.http}'"
  end

  # All Vagrant configuration is done here. The most common configuration
  # options are documented and commented below. For a complete reference,
  # please see the online documentation at vagrantup.com.

  # Every Vagrant virtual environment requires a box to build off of.
  box_name = 'company_trusty64_15052017.box'

  # We mirror the box images to most of the offices, so use the location nearest you.
  config.vm.box = "#{site_settings[:box_mirror]}/#{box_name}"

  # Disable automatic box update checking. If you disable this, then
  # boxes will only be checked for updates when the user runs
  # `vagrant box outdated`. This is not recommended.
  # config.vm.box_check_update = false

  # Create a forwarded port mapping which allows access to a specific port
  # within the machine from a port on the host machine. In the example below,
  # accessing 'localhost:8080' will access port 80 on the guest machine.
  # config.vm.network 'forwarded_port', guest: 80, host: 8080

  system_name = File.basename(File.dirname(__FILE__))

  case site
    when :vpn then begin
      # In the case of vpn connections, we will add a public adaptor, but will use the nat adaptor as the default route.  This should ensure that requests to the distirbution servers get routed over the vpn.
      config.vm.network 'public_network', type: 'dhcp'
      # TODO - THIS REMAINS TO BE TESTED.  If this doesn't work, remove the adaptor above, and uncomment the port forwards below.  You can then access via http://localhost:8080 etc.
      # config.vm.network "forwarded_port", guest: 8080, host: 8080
      # config.vm.network "forwarded_port", guest: 80, host: 8088 # port 80 is a privledged port, so we have to use an alternative. in this case 8088
    end
    else
      # For a dhcp bridged interface, configured as default route.
      config.vm.network 'public_network', type: 'dhcp', :use_dhcp_assigned_default_route => true
      # For a statically assigned interface.
      # config.vm.network 'public_network', ip: 'the ip you want to configure the VM to have'

      # Vagrant doesn't appear to directly support a default gateway for a public network using static IP.
      # Here we issue the commands "manually" to change the route, using a handy variable to make it easy to adjust.
      # This is only needed if you want to access your VM on a different network.  (e.g. VPNed in, another department accessing it) etc.

      # IP address of your LAN's router
      default_router = 'ip'
  end


  # Create a private network, which allows host-only access to the machine
  # using a specific IP.
  # config.vm.network 'private_network', ip: '192.168.33.10'

  # Create a public network, which generally matched to bridged network.
  # Bridged networks make the machine appear as another physical device on
  # your network.

  # change/ensure the default route via the local network's WAN router, useful for public_network/bridged mode
  # UNCOMMENT TO SET THE ROUTE
  #config.vm.provision :shell, :inline => "sudo ip route delete default 2>&1 >/dev/null || true; sudo ip route add default via #{default_router}"

  # If true, then any SSH connections made will enable agent forwarding.
  # Default value: false
  # config.ssh.forward_agent = true


  # Share an additional folder to the guest VM. The first argument is
  # the path on the host to the actual folder. The second argument is
  # the path on the guest to mount the folder. And the optional third
  # argument is a set of non-required options.
  # config.vm.synced_folder '../data', '/vagrant_data'


  # Provider-specific configuration so you can fine-tune various
  # backing providers for Vagrant. These expose provider-specific options.
  # Example for VirtualBox:
  #
  config.vm.provider 'virtualbox' do |vb|
    # Don't boot with headless mode
    # vb.gui = true

    # Use VBoxManage to customize the VM. For example to change memory:
    #vb.customize 'pre-boot', ['...']
    vb.customize ['modifyvm', :id, '--memory', '4096']
    vb.customize ['modifyvm', :id, '--cpus', '4']
    # If using NAT adapter for DNS, use the VM Host's resolver.  (Make company DNS lookups work when using VPN)
    vb.customize ['modifyvm', :id, '--natdnshostresolver1' , 'on']

    # Take the beginning of the name from the folder containing this Vagrantfile
    containing_folder = File.basename(File.dirname(__FILE__))
    vb.name = "#{containing_folder}  - Vagrant provisioned"
  end

  config.vm.provider :vsphere do |vsphere, override|
    override.vm.box = 'dzirkler/vsphere-dummy'
    override.ssh.username = 'company'
    override.ssh.password = 'company1'
    override.nfs.functional = false

    vsphere.host = vsphere_config['host']
    vsphere.data_center_name = vsphere_config['data_center']
    vsphere.compute_resource_name = vsphere_config['compute_resource']
    vsphere.data_store_name = vsphere_config['data_store']
    vsphere.template_name = vsphere_config['template']
    vsphere.vm_base_path = vsphere_config['vm_base_path']

    vsphere.user = vsphere_config['user']
    vsphere.password = vsphere_config['password']
    vsphere.insecure = vsphere_config['insecure']
  end

  # Now that we are running in minion/master mode of salt, it is necessary to reset the minion_id on the machine, and update the salt minion file.  To reset the minion_id you can use the command 'salt_reset_minion_id'
  config.vm.hostname = 'app-dev'

  #This next bit fixes the 'stdin is not a tty' error when shell provisioning Ubuntu boxes
  config.vm.provision :shell,
    #if there a line that only consists of 'mesg n' in /root/.profile, replace it with 'tty -s && mesg n'
    :inline => %q[(grep -q -E '^mesg n$' /root/.profile && sed -i 's/^mesg n$/tty -s \&\& mesg n/g' /root/.profile && echo 'Ignore the previous error about stdin not being a tty. Fixing it now...') || exit 0;]

  ## Application dev boxes have become RAM constrained so a swapfile of 4GB is now used
  config.vm.provision "shell", path: "provisioning_scripts/create_swap.sh"

  ## Add a sync folder to alllow us to exchange content.
  config.vm.synced_folder 'storage', '/srv/storage/'

  ## Mount the salt file root
  config.vm.synced_folder 'salt/roots/', '/srv/salt/'

  ## Use all the defaults:
  config.vm.provision :salt do |salt|
    # Config Files
    salt.minion_config = 'salt/minion'
    salt.master_config = 'salt/master'

    # Pre generating and distributing seeded keys.
    salt.minion_key = 'salt/key/minion.pem'
    salt.minion_pub = 'salt/key/minion.pub'
    salt.master_key = 'salt/key/master.pem'
    salt.master_pub = 'salt/key/master.pub'

    # Install a master on this machine
    salt.install_master = true
    # Pre-seed the master
    salt.seed_master = {config.vm.hostname => salt.minion_pub}

    # Actions
    # Normally we want to run state.highstate to provision the machine
    salt.run_highstate = true

    # this resolves a bug in Vagrant 1.7.4 which prevents salt from bootstrapping correctly.
    salt.bootstrap_options = '-F -c /tmp -P -D'
  end

  # We need to run the various manual provisioning steps we would normally have the projects team run ...
  [
    # Force install the desired salt version
    %q[sudo apt-get install -qq -o Dpkg::Options::="--force-confold" -y --force-yes salt-common=2016.11.4+ds-1 salt-master=2016.11.4+ds-1 salt-minion=2016.11.4+ds-1;],
    # Make sure Ruby is installed, it is not included on VMware images
    %q[sudo apt-get install -qq -y ruby;],
    # Other shell commands
  ].each {|cmd| config.vm.provision :shell, :inline => cmd}
  config.vm.provision 'shell', inline: %q[LANG=c ifconfig | grep -B1 "inet addr" |awk '{ if ( $1 == "inet" ) { print $2 } else if ( $2 == "Link" ) { printf "%s:" ,$1 } }' |awk -F: '{ print $1 ": " $3 }']
end
```

### Debug output
https://gist.github.com/amaclean199/d52529b3ff9cd12f619e59fbb936b998

### Expected behavior
Vagrant should have provisioned a VM in vSphere.

### Actual behavior
It stopped provisioning after the salt provisioner section and reported that the masterless setting shouldn't exist, but it is not in the Vagrantfile at all. Also `vagrant validate` reports that the Vagrantfile is okay.
```ruby
==> default: Running provisioner: shell...
There are errors in the configuration of this machine. Please fix
the following errors and try again:

salt provisioner:
* The following settings shouldn't exist: masterless?
```

### Steps to reproduce
1. Provide parameters to config file to deploy a box in vSphere
2. `vagrant up --provider vsphere`

### References
- GH-7934

**Comment by amaclean199** [comment]
I should also mention that `vagrant provision --provision-with salt` and `vagrant provision --provision-with shell` work fine separately.

**Comment by afischer-opentext-com** [comment]
I am running into the same issue when using the vagrant-reload plugin in combination with the salt provisioner. 

```
    server.vm.provision :salt do |salt|
      salt.log_level = 'all'
      salt.verbose = true
      salt.masterless = false
      salt.run_highstate = true
    end
    
    server.vm.provision :reload
```

When checking the code where this comes from I came across the situation that imho that config attribute is useless as there is also a masterless attribute which has the same purpose. Imho this is just syntactic sugar which can be removed. 

I changed the code accordingly [The diff here](https://github.com/afischer-opentext-com/vagrant/commit/0b87184ad42810c655d814c98fd0d6bdd73b23c1?diff). 

Maybe that's good enough for a Pull Request, but not finally sure.
